### PR TITLE
GO-2043: make import process not blocking

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### Description
1. Add isSync option to Import function, so for Gallery and Use cases this process will be synced and general import will be async
2. Add new abstraction NotificationProcess, which could send notifications after process complete
3. Send event with rootCollectionID instead of returning it to ObjectImport request

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Related Tickets & Documents

https://linear.app/anytype/issue/GO-2043/unify-logic-for-long-background-requests-from-client
### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [x] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [x] 🙅 no documentation needed

### CODE REVIEW

File: core/block/import/importer.go [hunk 1]
```diff
@@ -96,45 +98,73 @@ func (i *Import) Init(a *app.App) (err error) {
...
 	i.oc = creator.New(i.s, factory, store, relationSyncer, fileStore, spaceService, objectCreator)
 	i.fileSync = app.MustComponent[filesync.FileSync](a)
+	i.notificationService = app.MustComponent[notifications.Notifications](a)
+	i.eventSender = app.MustComponent[event.Sender](a)
 	return nil
 }
 
 // Import get snapshots from converter or external api and create smartblocks from them
-func (i *Import) Import(ctx context.Context, req *pb.RpcObjectImportRequest, origin model.ObjectOrigin, progress process.Progress) (string, string, error) {
+func (i *Import) Import(ctx context.Context,
```
Comment (deff7) [inline core/block/import/importer.go:107]: maybe it's time to use `type ImportRequest struct`?

Comment (AnastasiaShemyakinskaya) [reply core/block/import/importer.go]: Fixed

File: core/block/import/importer.go [hunk 2]
```diff
@@ -96,45 +106,64 @@ func (i *Import) Init(a *app.App) (err error) {
...
-	defer i.Unlock()
-	isNewProgress := false
-	if progress == nil {
-		progress = i.setupProgressBar(req)
-		isNewProgress = true
+func (i *Import) Import(ctx context.Context, importRequest *ImportRequest) (string, error) {
+	if importRequest.IsSync {
+		return i.importObjects(ctx, importRequest)
+	}
+	go func() {
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply core/block/import/importer.go]: Add it. Please take a look, if I did it right

File: util/conc/conc.go [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
+	"runtime/debug"
 	"sync"
 
 	"github.com/hashicorp/go-multierror"
 	"github.com/samber/lo/parallel"
+
+	"github.com/anyproto/anytype-heart/pkg/lib/logging"
 )
 
+var log = logging.Logger("anytype-mw-api")
```
Comment (deff7) [inline util/conc/conc.go:14]: fix logger name

**Review by deff7 — APPROVED** [review]

File: util/conc/conc.go [hunk 4]
```diff
@@ -29,3 +35,20 @@ func MapErr[T, R any](input []T, f func(T) (R, error)) ([]R, error) {
 
 	return res, allErrors
 }
+
+func Go(fn func()) {
```
Comment (deff7) [inline util/conc/conc.go:39]: maybe add `go` keyword inside?

File: core/block/import/importer.go [hunk 5]
```diff
@@ -100,60 +103,66 @@ func (i *Import) Init(a *app.App) (err error) {
...
-	origin objectorigin.ObjectOrigin,
-	progress process.Progress,
-) *ImportResponse {
-	if req.SpaceId == "" {
+func (i *Import) Import(ctx context.Context, importRequest *ImportRequest) *ImportResponse {
+	if importRequest.IsSync {
+		return i.importObjects(ctx, importRequest)
+	}
+	go conc.Go(func() {
+		res := i.importObjects(context.Background(), importRequest)
```
Comment (deff7) [inline core/block/import/importer.go:117]: I suggest adding component context to import and pass it here

File: core/block/import/importer.go [hunk 6]
```diff
@@ -100,60 +103,66 @@ func (i *Import) Init(a *app.App) (err error) {
...
 	defer func() {
-		i.finishImportProcess(returnedErr, progress)
-		i.sendFileEvents(returnedErr)
-		i.recordEvent(&metrics.ImportFinishedEvent{ID: importId, ImportType: req.Type.String()})
+		i.onImportFinish(res, importRequest, importId)
 	}()
-	if i.s != nil && !req.GetNoProgress() && isNewProgress {
-		i.s.ProcessAdd(progress)
+	if i.s != nil && !importRequest.GetNoProgress() && isNewProgress {
+		i.s.ProcessAdd(importRequest.Progress)
```
Comment (deff7) [inline core/block/import/importer.go:147]: check error

**Review by deff7 — APPROVED** [review]

**Pull request merged**